docs/architecture/adr-043-nft-module.md

```diff
@@ -0,0 +1,320 @@
+
+## Decision
+
+We will create a module `x/nft`, which contains the following functionality:
+
+- Store NFTs and track their ownership.
+- Expose `Keeper` interface for composing modules to mint and burn NFTs.
+- Expose external `Message` interface for users to transfer ownership of their NFTs.
+- Query NFTs and their supply information.
+
+### Types
+
+#### Class
+
+We define a model for NFT **Class**, which is comparable to an ERC721 Contract on Ethereum, under which a collection of NFTs can be created and managed.
+
+```protobuf
+message Class {
+  string id          = 1;
+  string name        = 2;
+  string symbol      = 3;
+  string description = 4;
+  string uri         = 5;
+}
+```
+
+- `id` is an alphanumeric identifier of the NFT class; it is used as the primary index for storing the class; _required_
+- `name` is a descriptive name of the NFT class; _optional_
+- `symbol` is the symbol usually shown on exchanges for the NFT class; _optional_
+- `description` is a detailed description of the NFT class; _optional_
+- `uri` is a URL pointing to an off-chain JSON file that contains metadata about this NFT class ([OpenSea example](https://docs.opensea.io/docs/contract-level-metadata)); _optional_
+
+#### NFT
+
+We define a general model for `NFT` as follows.
+
+```protobuf
+message NFT {
+  string class_id           = 1;
+  string id                 = 2;
+  string uri                = 3;
+  google.protobuf.Any data  = 10;
+}
+```
+
+- `class_id` is the identifier of the NFT class where the NFT belongs; _required_
+- `id` is an alphanumeric identifier of the NFT, unique within the scope of its class. It is specified by the creator of the NFT and may be expanded to use DID in the future. `class_id` combined with `id` uniquely identifies an NFT and is used as the primary index for storing the NFT; _required_
+  ```
+  {class_id}/{id} --> NFT (bytes)
+  ```
+- `uri` is a URL pointing to an off-chain JSON file that contains metadata about this NFT (Ref: [ERC721 standard and OpenSea extension](https://docs.opensea.io/docs/metadata-standards)); _required_
+- `data` is a field that CAN be used by composing modules to specify additional properties for the NFT; _optional_
+
+This ADR doesn't specify values that `data` can take; however, best practices recommend upper-level NFT modules clearly specify their contents.  Although the value of this field doesn't provide the additional context required to manage NFT records, which means that the field can technically be removed from the specification, the field's existence allows basic informational/UI functionality.
+
+### `Keeper` Interface
+
+```go
+type Keeper interface {
+  NewClass(class Class)
+  UpdateClass(class Class)
+
+  Mint(nft NFT，receiver sdk.AccAddress)   // updates totalSupply
+  Burn(classId string, nftId string)    // updates totalSupply
+  Update(nft NFT)
```

Comment:
```suggestion
  Update(nft NFT)
```

We need to be very careful about how this `Update` method is used if ICS30 transfers metadata. Basically once an NFT is escrowed, metadata would be locked. If there is a module who's logic allows a minter (not owner) of an NFT to modify the metadata, that could result in an inconsistent state.